src/stan_math_backend/Lower_stmt.ml

```diff
@@ -174,45 +230,48 @@ let rec lower_statement Stmt.Fixed.{pattern; meta} : stmt list =
   @
   match pattern with
   | Assignment
-      ( (vident, _, [])
-      , ( {pattern= FunApp (CompilerInternal (FnReadData | FnReadParam _), _); _}
-        as rhs ) ) ->
-      Assign (Var vident, lower_expr rhs) |> wrap_e
+      ( ((LVariable _ | LTupleProjection _ | LIndexed (_, [])) as lhs)
```

Comment:
How much more annoying does everything else become if we were to eliminate `LIndexed` in favor of the above `lvalue/lbase` type pair where empty index list represents unindexed lvalues? I don't immediately see why it would be a problem but I haven't looked that closely.

Reply:
I think it wouldn't be that different from what we have right now in the MIR (`and 'a lvalue = string * UnsizedType.t * 'a Index.t list`), but I also like giving the indexing an explicit name, since we already do in the AST.

Put another way, I'm happier having this weird edge case of `LIndexed []` compared to the other method, where a index list of `[]` is actually very important since that is what makes something a variable as opposed to an indexed lhs.

The next version of `Core_kernel` after the one we're currently using has a `Nonempty_list` module which seems like it might be a reasonably ergonomic solution down the road.

Reply:
Don't we now also have `LIndexed (LIndexed (_, _), _)` as a weird edge case?

Reply:
Fair enough. I'll look into how bad it would be to change. This is the thing that required the most changes to existing code in the optimization pipeline so I'm hesitant, but it's worth exploring.

Reply:
This didn't end up being too bad and I like some of the things like this that it cleaned up. I think we can do a similar refactor to the AST type, but it needs to wait for the nested multi-indexing lvalues deprecation?